src/composer.rs

```diff
@@ -0,0 +1,1052 @@
+
+        witness
+    }
+
+    /// Asserts `a == b` by appending a gate
+    fn assert_equal(&mut self, a: Witness, b: Witness) {
+        let constraint =
+            Constraint::new().left(1).right(-BlsScalar::one()).a(a).b(b);
+
+        self.append_gate(constraint);
+    }
+
+    /// Adds a logical AND gate that performs the bitwise AND between two values
+    /// for the specified first `num_bits` returning a [`Witness`]
+    /// holding the result.
+    ///
+    /// # Panics
+    ///
+    /// If the `num_bits` specified in the fn params is odd.
+    fn append_logic_and(
+        &mut self,
+        a: Witness,
+        b: Witness,
+        num_bits: usize,
+    ) -> Witness {
+        self.append_logic_component(a, b, num_bits, false)
+    }
+
+    /// Adds a logical XOR gate that performs the XOR between two values for the
+    /// specified first `num_bits` returning a [`Witness`] holding the
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// If the `num_bits` specified in the fn params is odd.
+    fn append_logic_xor(
+        &mut self,
+        a: Witness,
+        b: Witness,
+        num_bits: usize,
+    ) -> Witness {
+        self.append_logic_component(a, b, num_bits, true)
+    }
+
+    /// Constrain `a` to be equal to `constant + pi`.
+    ///
+    /// `constant` will be defined as part of the public circuit description.
+    fn assert_equal_constant<C: Into<BlsScalar>>(
+        &mut self,
+        a: Witness,
+        constant: C,
+        public: Option<BlsScalar>,
+    ) {
+        let constant = constant.into();
+        let constraint = Constraint::new().left(1).constant(-constant).a(a);
+        let constraint =
+            public.map(|p| constraint.public(p)).unwrap_or(constraint);
+
+        self.append_gate(constraint);
+    }
+
+    /// Asserts `a == b` by appending two gates
+    fn assert_equal_point(&mut self, a: WitnessPoint, b: WitnessPoint) {
+        self.assert_equal(*a.x(), *b.x());
+        self.assert_equal(*b.y(), *b.y());
+    }
+
+    /// Asserts `point == public`.
+    ///
+    /// Will add `public` affine coordinates `(x,y)` as public inputs
+    fn assert_equal_public_point<P: Into<JubJubAffine>>(
+        &mut self,
+        point: WitnessPoint,
+        public: P,
+    ) {
+        let public = public.into();
+
+        self.assert_equal_constant(
+            *point.x(),
+            BlsScalar::zero(),
+            Some(-public.get_x()),
+        );
+
+        self.assert_equal_constant(
+            *point.y(),
+            BlsScalar::zero(),
+            Some(-public.get_y()),
+        );
+    }
+
+    /// Adds two curve points by consuming 2 gates.
+    fn component_add_point(
+        &mut self,
+        a: WitnessPoint,
+        b: WitnessPoint,
+    ) -> WitnessPoint {
+        // In order to verify that two points were correctly added
+        // without going over a degree 4 polynomial, we will need
+        // x_1, y_1, x_2, y_2
+        // x_3, y_3, x_1 * y_2
```

Comment:
This is extended from the previous documentation. Since its not the scope of this PR, I'm opening an issue to better document that function in a different one

https://github.com/dusk-network/plonk/issues/703